doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +962,563 @@ tactic
-   :tacn:`finish_timing` at each level of nesting.
+By copying the definition of :tacn:`time_constr` from the standard library,
+users can achieve support for a fixed pattern of nesting by passing
+different :token:`string` parameters to :tacn:`restart_timer` and
+:tacn:`finish_timing` at each level of nesting.
 
-   .. example::
+.. example::
 
-      .. coqtop:: all abort
+   .. coqtop:: all abort
 
-         Ltac time_constr1 tac :=
-           let eval_early := match goal with _ => restart_timer "(depth 1)" end in
-           let ret := tac () in
-           let eval_early := match goal with _ => finish_timing ( "Tactic evaluation" ) "(depth 1)" end in
-           ret.
+      Ltac time_constr1 tac :=
+        let eval_early := match goal with _ => restart_timer "(depth 1)" end in
+        let ret := tac () in
+        let eval_early := match goal with _ => finish_timing ( "Tactic evaluation" ) "(depth 1)" end in
+        ret.
+
+      Goal True.
+        let v := time_constr
+             ltac:(fun _ =>
+                     let x := time_constr1 ltac:(fun _ => constr:(10 * 10)) in
+                     let y := time_constr1 ltac:(fun _ => eval compute in x) in
+                     y) in
+        pose v.
+
+Local definitions: let
+~~~~~~~~~~~~~~~~~~~~~~
 
-         Goal True.
-           let v := time_constr
-                ltac:(fun _ =>
-                        let x := time_constr1 ltac:(fun _ => constr:(10 * 10)) in
-                        let y := time_constr1 ltac:(fun _ => eval compute in x) in
-                        y) in
-           pose v.
-
-Local definitions
-~~~~~~~~~~~~~~~~~
+.. tacn:: let {? rec } @let_clause {* with @let_clause } in @ltac_expr
 
-Local definitions can be done as follows:
+   .. insertprodn let_clause let_clause
 
-.. tacn:: let @ident__1 := @ltac_expr__1 {* with @ident__i := @ltac_expr__i} in @ltac_expr
-   :name: let ... := ...
+   .. prodn::
+      let_clause ::= @name := @ltac_expr
+      | @ident {+ @name } := @ltac_expr
 
-   each :n:`@ltac_expr__i` is evaluated to :n:`v__i`, then, :n:`@ltac_expr` is evaluated
-   by substituting :n:`v__i` to each occurrence of :n:`@ident__i`, for
-   i = 1, ..., n. There are no dependencies between the :n:`@ltac_expr__i` and the
-   :n:`@ident__i`.
+   Evaluates each :n:`@let_clause`, substitutes the bound variables into
+   :n:`@ltac_expr` and then evaluates :n:`@ltac_expr`.  There are no dependencies
+   between the :n:`@let_clause`\s.
 
-   Local definitions can be made recursive by using :n:`let rec` instead of :n:`let`.
-   In this latter case, the definitions are evaluated lazily so that the rec
-   keyword can be used also in non-recursive cases so as to avoid the eager
+   Creates a local definition.
+   Using :tacn:`let` `rec` makes the definition recursive and causes
+   the definitions to be evaluated lazily.  Thus, :tacn:`let` `rec`
+   can be used in non-recursive cases to avoid eager
```

Comment:
> Doesn't it apply only in case `rec` was not provided?

Indeed,
```coq
Goal True.
  let x _ := y ()
  with y _ := x () in
  idtac.
```
fails but
```coq
Goal True.
  let rec x _ := y ()
  with y _ := x () in
  idtac.
```
succeeds.

Reply:
What does (as I reworded it) "`rec` can also be used for non-recursive cases to avoid eager evaluation of local definitions." mean?  How is it useful?  @JasonGross Can you give an example?

Reply:
```coq
Goal True /\ True.
  Fail let x := match goal with |- True => idtac end in
       split; x. (* Error: No matching clauses for match. *)
  let rec x := match goal with |- True => idtac end in
  split; x. (* success *)
```
Doing `let rec f := body in` for a non-recursive body seems to have exactly the same effect as as prefixing `body` with `idtac; `.  (Note that this trick is useless for local definitions that evaluate to something other than a tactic, because, I believe, without arguments, there's no way to get the resulting thunked tactic to actually run.)

Reply:
Then I think it's better to omit it.  It would need more explanation to make it clear and the benefit is quite small.  It seems like an artifact of how things happen to be implemented rather than intentional behavior.  It's possible that this will change in Ltac or Ltac2 if the implementation changes for other reasons.  Documenting it moves it closer to being a promised behavior.

Reply:
Ltac2 does not have this same separation between tactics effecting the goal vs tactics returning a value, and does not have this same two-phase evaluation semantics, so this bit will not apply to Ltac2.  I don't see things around this changing in Ltac1, though I suppose it is possible.  I don't see a real harm in documenting it here, but it also seems fine to omit this corner-case.  (I would lean towards documenting it just for completeness' sake, and Ltac is already filled with accidental artifacts that have become relied-upon features, but I'm okay with it either way.)